### DISCUSSION TITLE
Multiple Root Layouts are messed up in production.

### DISCUSSION THREAD

**Discussion by wkd-kapsule** [vercel/next.js#60078] [Help]
### Summary

Hi everyone, I hope you're having a great day.

Mine was doing pretty fine until I stumbled upon an issue I can't seem to figure out... I am building a multi-tenant app. Its users will be able to create their own website which will be hosted as a subdomain of my app until they add a custom domain.

I redirect "localhost:3000/domain/mysite" to "mysite.localhost:3000" using the middleware. I check if the pathname starts with "/domain".

And following the guidelines from the docs, I use multiple Root Layouts with Route Groups to display a totally different Layout on all localhost:3000/domain/* pages. It works fine in development: when I'm at "localhost:3000" (home), I get the corresponding Root Layout. Then when I go to "localhost:3000/domain/test" I get redirected to "test.localhost:3000" with the different Root Layout.

But in production, any route just shows the same page: the home (page.tsx) that is supposed to be shown only on *.app.com routes. And the redirection from app.com/domain/mysite to mysite.app.com isn't even working like in dev. The url just stays as app.com/domain/mysite and the page shown is still the same one that is shown no matter what route I visit.

Here are my folders structure, middleware function and some screenshots of the app:

![files](https://github.com/vercel/next.js/assets/77295465/15963e2f-800d-4992-b32a-cc9658543bd6)

![var](https://github.com/vercel/next.js/assets/77295465/a4537ba5-3b83-41f0-9504-93bfaf7dd989)

![middleware](https://github.com/vercel/next.js/assets/77295465/22eb8be6-7614-4e08-a048-7513dca98922)

What my home page looks like on localhost:3000
![image](https://github.com/vercel/next.js/assets/77295465/2676a1db-4b5e-4359-a693-ada9348b0c0d)

localhost:3000/domain/custom redirected to custom.localhost:3000
![image](https://github.com/vercel/next.js/assets/77295465/39bc9937-e26f-415b-a0a9-312b79a0a017)

Production site on home page (not subdomain)
![image](https://github.com/vercel/next.js/assets/77295465/e7c09c8c-4430-461f-b8cc-b00cae8ffbcb)

app.com/domain/custom... Site behaves like it just skips the middleware redirect
![image](https://github.com/vercel/next.js/assets/77295465/679a1e1d-395c-489f-b614-cebae882f3e1)

When I manually type the subdomain in the url and visit the page
![image](https://github.com/vercel/next.js/assets/77295465/6f055411-2353-4c0f-a76f-017adc3a282b)

As you can see, it works perfectly fine in dev but for some reason, the production build seems to be stuck on that secondary layout that should be displayed only on specific routes. If it wasn't enough, it looks like the middleware is ignored in production.

Some info about my domain: I have purchased it on namecheap and added it to Vercel. It is correctly configured with the subdomain wildcard (otherwise [custom.app.com](https://custom.app.com/) would just throw an error).

To make it even worse. I tried to revert the repo to what is was before the second Root Layout. I've removed the (sites) routes group and reverted my middleware as well. And now guess what? The localhost is still working as it should and production shows a 404 on all routes. It hints me that somehow my original Root Layout is just ignored no matter what. So much so that when I revert to a single one (that one), the production build doesn't find anything to display.

It is very weird because when I delete the layout.tsx file and try to deploy, Vercel complains that the file is missing. But then proceeds to ignore it all together when it's there...

Please, any help would be very very much appreciated...

**Comment by issam-hub** [comment]
one question please, where is your middleware file located ?
